pkg/ccl/sqlproxyccl/acl/watcher.go

```diff
@@ -173,14 +160,32 @@ func (l *listener) Less(than btree.Item) bool {
 	return l.id < than.(*listener).id
 }
 
-func checkConnection(connection ConnectionTags, list *Denylist) error {
-	ip := DenyEntity{Item: connection.IP, Type: IPAddrType}
-	if err := list.Denied(ip); err != nil {
-		return errors.Wrapf(err, "connection ip '%v' denied", connection.IP)
-	}
-	cluster := DenyEntity{Item: connection.Cluster, Type: ClusterType}
-	if err := list.Denied(cluster); err != nil {
-		return errors.Wrapf(err, "connection cluster '%v' denied", connection.Cluster)
+func checkListeners(
+	listeners *btree.BTree, timesource timeutil.TimeSource, controllers ...AccessController,
+) {
+	listeners.Ascend(func(i btree.Item) bool {
+		lst := i.(*listener)
+		if err := checkConnection(lst.connection, timesource, controllers...); err != nil {
+			lst.mu.Lock()
+			defer lst.mu.Unlock()
+			if lst.mu.denied != nil {
+				lst.mu.denied(err)
+				lst.mu.denied = nil
+			}
+		}
+		return true
+	})
+}
+
+func checkConnection(
+	connection ConnectionTags, timesource timeutil.TimeSource, controllers ...AccessController,
+) error {
+	for _, c := range controllers {
+		if c != nil {
```

Comment:
In the case where there is an error reading the files on startup, it will return a `nil` AccessController. But if we're going to fail to start in that case, we can remove this check.